[laurenlim2112] iP

laurenlim2112 opened the pull request:
## Spot
> Everything I know I learned from dogs.

Who's Spot? Well, he's:

- A very cute puppy 🐕 
- **Super** cute
- Did I mention *really* cute ?

What can Spot do?
1. Keep track of your tasks
2. Be cute
3. Help find more tasks for you...wait...

He ticks all the boxes you're looking for in a chatbot!
- [x] being cute
- [x] being adorable
- [x] keeping track of your tasks I guess

If you're a Java programmer, you can practise your Java skills with Spot too. Here's the `main` method:

``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Want to know how this description was made? Click [here](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) to find out!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

laurenlim2112 committed:
Duke.java: rename chatbot, implement skeletal version

laurenlim2112 committed:
Duke.java: echoes user commmands, exits on 'bye' command

laurenlim2112 committed:
Duke.java: store and list tasks added by user

laurenlim2112 committed:
Add Task class to represent tasks and ability to mark as done/not done

laurenlim2112 committed:
Support tracking 3 task types

laurenlim2112 committed:
Test using the I/O redirection technique

laurenlim2112 committed:
Add error handling

laurenlim2112 committed:
Add support for task deletion

laurenlim2112 committed:
Further improve error handling

laurenlim2112 committed:
Implement Level 7: Save

laurenlim2112 committed:
Merge branch 'branch-Level-7'

laurenlim2112 committed:
Add the ability to understand dates

laurenlim2112 committed:
Merge branch 'branch-Level-8'

laurenlim2112 committed:
Extract closely-related code as classes

laurenlim2112 committed:
Merge branch 'branch-A-MoreOOP'

laurenlim2112 committed:
Divide classes into packages

laurenlim2112 committed:
Merge branch 'branch-A-Packages'

laurenlim2112 committed:
Merge branch 'add-gradle-support'

laurenlim2112 committed:
Test commit

laurenlim2112 committed:
Merge branch 'master' of https://github.com/laurenlim2112/ip

# Conflicts:
#	gradlew
#	gradlew.bat

laurenlim2112 committed:
Edit gitignore

laurenlim2112 committed:
Add JUnit tests to test the behaviour of the code.

laurenlim2112 committed:
Merge branch 'branch-A-JUnit'

laurenlim2112 committed:
Package the app as an executable JAR file

laurenlim2112 committed:
Add JavaDoc comments to the code

laurenlim2112 committed:
Add configuration files for checkstyle

laurenlim2112 committed:
Tweak the code to comply with the given coding standard

laurenlim2112 committed:
Add task-searching feature

laurenlim2112 committed:
Merge branch 'branch-A-JavaDoc'

laurenlim2112 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/spot/Storage.java
#	src/main/java/spot/exception/SpotException.java

laurenlim2112 committed:
Merge branch 'A-Level-9'

laurenlim2112 committed:
Further tweak code to comply with coding standards

laurenlim2112 committed:
Add a GUI to the chatbot using JavaFX

laurenlim2112 committed:
Merge branch 'branch-Level-10'

laurenlim2112 committed:
Fix coding style violations

jonyeokj commented on src/main/java/spot/Main.java:
```diff
@@ -0,0 +1,41 @@
+            });
+            stage.setTitle("Spot");
+            stage.show();
+        } catch (IOException e) {
```
In the start method, when catching the IOException, only the stack trace is printed. Consider adding a more descriptive error message to inform the user about the error. For example:

```java
} catch (IOException e) {
    System.err.println("Failed to load the main window.");
    e.printStackTrace();
}
```

jonyeokj commented on src/main/java/spot/Main.java:
The Main class name is a bit generic. Consider a more descriptive name like SpotApp or Spot. Isn't strictly against the conventions, but a nit and suggestion for clarity. 👍

jonyeokj commented on src/main/java/spot/Main.java:
```diff
@@ -0,0 +1,41 @@
+    public void start(Stage stage) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
+            AnchorPane ap = fxmlLoader.load();
```
Just a nit, the variable name ap is a bit too short and not descriptive. A more descriptive name like anchorPane. Rest of the code LGTM.

jonyeokj commented on src/main/java/spot/MainWindow.java:
```diff
@@ -0,0 +1,79 @@
+        String input = userInput.getText();
+        String response = spot.getResponse(input);
+        dialogContainer.getChildren().addAll(
+                DialogBox.getUserDialog(input, userImage),
```
Good practice to handle potential NullPointerException that might be thrown if the resource path for userImage is not found. Consider a try-catch or an assertion to check if the resource is non-null.

jonyeokj commented on src/main/java/spot/MainWindow.java:
```diff
@@ -0,0 +1,79 @@
+        String response = spot.getResponse(input);
+        dialogContainer.getChildren().addAll(
+                DialogBox.getUserDialog(input, userImage),
+                DialogBox.getSpotDialog(response, spotImage)
```
Same for spotImage if the resource path is not found.

jonyeokj commented on src/main/java/spot/Spot.java:
```diff
@@ -0,0 +1,52 @@
+package spot;
+
```
Excessive whitespace between import and package statements

jonyeokj commented on src/main/java/spot/Parser.java:
```diff
@@ -0,0 +1,134 @@
+import spot.command.MarkCommand;
+import spot.command.UnmarkCommand;
+import spot.exception.SpotException;
+
```
Excessive whitespace after import statements.

jonyeokj commented on src/main/java/spot/Parser.java:
```diff
@@ -0,0 +1,134 @@
+     * @return Resulting command.
+     * @throws SpotException  If input string is invalid.
+     */
+    public static Command parseCommand(String input) throws SpotException {
```
parseCommand is quite long and does a lot of things, might be beneficial to break it down or add in-line comments to add clarity.

jonyeokj commented on src/main/java/spot/Parser.java:
```diff
@@ -0,0 +1,134 @@
+            return new UnmarkCommand(position);
+        } else if (input.startsWith("delete")) {
+            if (input.length() <= 7) {
+                throw new SpotException("Spot needs more details than that!");
```
Consider constants for repeated strings. For instance, the error message "Spot needs more details than that!" appears multiple times. Can help with consistency, just a nit.

jonyeokj reviewed (commented):
The code is modular with clear separations of concerns, helps with code design and maintainability in the future. Naming conventions are largely consistent and clear. Error handling is also done well. Appropriate javadocs also observed to help understand the code.

parseCommand could be refactored to be smaller methods as it is extremely lengthy.

LGTM overall!

laurenlim2112 committed:
Use asserts to document assumptions that should hold in the code

Various assumptions are made at several points in the code.

However, these assumptions may not always hold.

Let's use Java assertions to check if these assumptions do indeed hold.

Their impact on performance is worth the additional safety provided.

laurenlim2112 committed:
Merge pull request #1 from laurenlim2112/branch-A-Assertions

Use asserts to document assumptions that should hold in the code

laurenlim2112 committed:
Improve the code quality where necessary

There are several areas where code quality is low, such as long methods.

This decreases readability of code.

Let's refactor the code where necessary to improve overall code quality.

Refactoring will improve the program's internal structure
in small steps, without affecting its behaviour.

laurenlim2112 committed:
Merge pull request #2 from laurenlim2112/branch-A-CodeQuality

Improve the code quality where necessary

laurenlim2112 committed:
Add support for editing details of tasks

laurenlim2112 committed:
Merge branch 'branch-BCD-Extension'

laurenlim2112 committed:
Improve readability of Parser code

laurenlim2112 committed:
Reduce repetitiveness of code in task-listing methods in TaskList

laurenlim2112 committed:
Add JavaDocs as needed

laurenlim2112 committed:
Add a screenshot of the UI

laurenlim2112 committed:
Add a User Guide

laurenlim2112 committed:
Update JUnit tests to reflect new error messages

laurenlim2112 committed:
Update User Guide
